### ISSUE TITLE
Uploading files as guest via voucher fails with: terasender_rest_xsrf_token_did_not_match

### ISSUE THREAD

**Issue by moschlar** [filesender/filesender#1305]
The log says:

```
[2022-08-08 14:11:43] [rest:info] authentication::create(1) saml_uid schlarbm@uni-mainz.de
[2022-08-08 14:11:43] [rest:info] authentication::create(2) FOUND AND RETURNING 15
[2022-08-08 14:11:44] [user 15] [rest:error] SECURITY OWASP CSRF protector log message: {"HOST":"...","REQUEST_URI":"\/filesender3\/rest.php\/file\/51\/chunk\/0?key=3acaa47a-8897-4b0b-bdb6-502504fa57fe&vid=254ab87d-c119-4c83-aaf6-203802b43638&roundtriptoken=3d2b6c2d973d5cf539e32a7a7d3f7409","requestType":"POST","cookie":{"csrfptoken":"61fd0ca989","PHPSESSID":"688qutilua9bm8o3elstf810jh","SimpleSAML":"ukqcfrbsiibr5luu3kog6e2ig3"},"timestamp":1659960704,"message":"OWASP CSRF PROTECTOR VALIDATION FAILURE"}
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException exception uid:62f0fd80575e5] rest_xsrf_token_did_not_match
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException trace uid:62f0fd80575e5] #0 /opt/filesender/filesender3/lib/vendor/owasp/csrf-protector-php/libs/csrf/csrfprotector.php(343): FileSendercsrfProtectorErrorHandler->failedValidationAction()
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException trace uid:62f0fd80575e5] #1 /opt/filesender/filesender3/lib/vendor/owasp/csrf-protector-php/libs/csrf/csrfprotector.php(246): csrfProtector::failedValidationAction()
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException trace uid:62f0fd80575e5] #2 /opt/filesender/filesender3/lib/vendor/owasp/csrf-protector-php/libs/csrf/csrfprotector.php(199): csrfProtector::authorizePost()
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException trace uid:62f0fd80575e5] #3 /opt/filesender/filesender3/classes/utils/Security.class.php(117): csrfProtector::init()
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException trace uid:62f0fd80575e5] #4 /opt/filesender/filesender3/classes/rest/RestServer.class.php(190): Security::validateAgainstCSRF()
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException trace uid:62f0fd80575e5] #5 /opt/filesender/filesender3/www/rest.php(37): RestServer::process()
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException trace uid:62f0fd80575e5] #6 {main}
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException details uid:62f0fd80575e5] method = "put"
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException details uid:62f0fd80575e5] endpoint = "file"
[2022-08-08 14:11:44] [user 15] [rest:error] [RestException details uid:62f0fd80575e5] "owasp tokens, see log"
```

**Comment by moschlar** [comment]
In my browser toolbox, I can see that I have *two* cookies named `csrfptoken` - one with the path `/filesender3` and one with the path `/filesender3/rest.php`. Maybe that is an additional hint?

**Comment by monkeyiq** [comment]
Is this with version 2.34? Firefox or Chrome?

**Comment by moschlar** [comment]
Yes, 2.34 and 3.0.alpha6.

And it happens in both Firefox and Chrome.

I'd be happy to do some debugging, if you can give me some hints what to try.

**Comment by monkeyiq** [comment]
There was some stuff enabled in 2.33 which may be causing this. "Stronger checks on guest vid, round trip token checks are forced on for guest uploads". I have just updated recently from Fedora 34 to 36 on the dev machine. I just used `development` which should be very close to 2.34 to invite myself and upload a file. 

One thought up front, do you have config/csrf-protector-config.php from your FileSender sources installed in `/opt/filesender/filesender3/config/`.  This might not make a difference, but it might and is worth a check first in case it does.

In the Chome devtools I see one `csrfptoken` sent for the `0?key=` request. There is also a `X-Filesender-Security-Token` but that is another check. 

The `0?key=` request should be after a number of js requests and starting with a `transfer?vid=` more toward the top of the action when you click send. For me the `transfer?vid=` sends the same token as `0?key=`.

Another idea is to take a look (in chome) in the devtools under the application tab. In the list on the left there is storage/cookies which you might clear out for the /filesender paths. This shouldn't really be an issue as those should be overridden when you start the session.

For my guest user session (a separate user login on my machine) I see a /filesender and /filesender/rest.php `csrfptoken`. I suspect there it doesn't matter too much that there are two as long as each path sees the token it expects to see.

It might be interesting to see what authorizePost() in csrfprotector.php is seeing for `$token` and `$_SESSION[self::$config['CSRFP_TOKEN']]` to see if a token is not getting through or if it is an unexpected value.

**Comment by monkeyiq** [comment]
I just moved my config/csrf-protector-config.php away and it gives a different more brutal error so that should not be this issue.

**Comment by monkeyiq** [comment]
if you look at csrfprotector.php / logCSRFattack() and add a line like the following you will be able to see a list of tokens in your log that are considered valid for the request:
```
            $context['expected'] = $_SESSION[self::$config['CSRFP_TOKEN']];
```
I also added this
```
            $context['expectedtokenname'] = self::$config['CSRFP_TOKEN'];
```
to make sure that "expectedtokenname":"csrfptoken" was in the logs.

**Comment by moschlar** [comment]
> The `0?key=` request should be after a number of js requests and starting with a `transfer?vid=` more toward the top of the action when you click send. For me the `transfer?vid=` sends the same token as `0?key=`.

The `csrfptoken` *sent* in both these requests is the same, but the first _response_ (for the request to `transfer?vid=`) is setting a new (different) token.
However, all requests still send the _initial_ token - I suspect the failure comes from the fact that it expects to see the _newer_ token, but that doesn't get sent at all.

The _initial_ token is stored for the `/filesender3` path and the _newer_ token for  `/filesender3/rest.php`.

**Comment by moschlar** [comment]
> if you look at csrfprotector.php / logCSRFattack() and add a line like the following you will be able to see a list of tokens in your log that are considered valid for the request:

Oh, wow, okay - so according to the log, it *should* work:

```
[2022-08-09 14:00:55] [user 15] [rest:error] SECURITY OWASP CSRF protector log message: {"HOST":"filesender.zdv.net","REQUEST_URI":"\/filesender3\/rest.php\/file\/56\/chunk\/0?key=30861b0a-b202-4fd9-bc13-67216247a37c&vid=254ab87d-c119-4c83-aaf6-203802b43638&roundtriptoken=8a221860363644f79486ec589bd2dcce","requestType":"POST","cookie":{"csrfptoken":"0ad2026cb3","PHPSESSID":"ci4ugac7s24kjire08tahvl1gl","SimpleSAML":"r7nf53isi6igsccrfcrrv8fftq"},"expected":["c50e400aec","0ad2026cb3"],"expectedtokenname":"csrfptoken","timestamp":1660046455,"message":"OWASP CSRF PROTECTOR VALIDATION FAILURE"}
[2022-08-09 14:00:58] [user 15] [rest:error] SECURITY OWASP CSRF protector log message: {"HOST":"filesender.zdv.net","REQUEST_URI":"\/filesender3\/rest.php\/transfer\/44?key=30861b0a-b202-4fd9-bc13-67216247a37c&vid=254ab87d-c119-4c83-aaf6-203802b43638&roundtriptoken=8a221860363644f79486ec589bd2dcce","requestType":"POST","cookie":{"csrfptoken":"0ad2026cb3","PHPSESSID":"ci4ugac7s24kjire08tahvl1gl","SimpleSAML":"r7nf53isi6igsccrfcrrv8fftq"},"expected":["c50e400aec","0ad2026cb3"],"expectedtokenname":"csrfptoken","timestamp":1660046458,"message":"OWASP CSRF PROTECTOR VALIDATION FAILURE"}
```

**Comment by monkeyiq** [comment]
Strange. I also see a list for expected and the presented token is in it so all is good for my testing.

It seems that one of the 3 items in authorizePost() isn't working the same way here and there. Since we can be fairly sure that the first two pass it is something in isValidToken()
```
                //currently for same origin only
                if (!($token && isset($_SESSION[self::$config['CSRFP_TOKEN']])
                    && (self::isValidToken($token)))) {
```

**Comment by moschlar** [comment]
Could it be that we are seeing the log messages for another (unproblematic) request because the function `authorizePost()` is handling POST/PUT/DELETE differently by getting the token from the form payload and not from the query string?

**Comment by moschlar** [comment]
I might be right - in https://github.com/filesender/filesender/blob/development/lib/vendor/owasp/csrf-protector-php/libs/csrf/csrfprotector.php#L239, $token is empty!

**Comment by monkeyiq** [comment]
I think the cookie and token it is using to compare are different. In logCSRFattack() I added this to see the token it is actually using:
```
            $context['token'] = self::getTokenFromRequest();
```

**Comment by monkeyiq** [comment]
Seems we both bumped into the same path at the same time :) That getTokenFromRequest() is not returning data there.

**Comment by moschlar** [comment]
Yep ;-)

Because it reads from `$_POST`, but we are in a PUT request?

**Comment by monkeyiq** [comment]
it looks like the following line of getTokenFromRequest() is what is returning the token here.
```
return $_SERVER[self::$tokenHeaderKey];
```

**Comment by monkeyiq** [comment]
and 'HTTP_CSRFPTOKEN' == self::$tokenHeaderKey

**Comment by monkeyiq** [comment]
So this should be coming from the csrftoken sent in the http request. Perhaps something is blocking that http header from reaching the code?

**Comment by monkeyiq** [comment]
fwiw I am on the following. Though it might be a config thing rather than a version thing if the header isn't making it through.
```
httpd-2.4.54-3.fc36.x86_64
php-8.1.8-1.fc36.x86_64
```

**Comment by moschlar** [comment]
I'm on Debian 11

apache2        2.4.54-1~deb11u1
php7.4-fpm     7.4.30-1+deb11u1

I remember something regarding HTTP headers getting strange names this way.
Let me see if I can dig that up.

**Comment by moschlar** [comment]
Hm, no, thats not it.
The variable is set in the `$_SERVER` array, but empty...

**Comment by monkeyiq** [comment]
So it is sent ok in the browser network log but doesn't arrive in the $_SERVER array. 

Is there anything that might be doing some ACL type stuff on http headers that make it into $_SERVER

**Comment by Erik-Horn** [comment]
I encountered the same issue with my recently upgraded filesender install. After doing a bunch of troubleshooting, I found that adding the following to my config.php file fixed the issue:

$config['owasp_csrf_protector_enabled']=true;

I don't know if this is the best way to fix it.


Here's some of the info I gathered while tracking down the problem:

Changing the setting to false causes the same failure. I'm guessing that the client-side code honors the setting, but the server side always assumes it's enabled. The default appears to be unset, which evaluates to false on the client-side.

The reason that the upload was failing with rest.php is that the csrfptoken http header was present but empty. This is caused by the check in getCSRFToken client.js:64. The configuration comes from filesender-config.js.php which had owasp_csrf_protector_enabled set to blank.

I guessed at adding it to the config.php and it happened to work.

If it matters, I'm running centos 7 with php-7.4.30-1.el7.remi.x86_64

**Comment by monkeyiq** [comment]
I have made https://github.com/filesender/filesender/pull/1306/files which allows guest uploads to happen regardless of if the owasp config is true or false. I will likely merge and release with that soon.

$config['owasp_csrf_protector_enabled']=false;

**Comment by monkeyiq** [comment]
This should be fixed in 2.35.
